packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m

```diff
@@ -234,9 +243,14 @@ - (AVMutableVideoComposition *)getVideoCompositionWithTransform:(CGAffineTransfo
   }
   videoComposition.renderSize = CGSizeMake(width, height);
 
-  // TODO(@recastrodiaz): should we use videoTrack.nominalFrameRate ?
-  // Currently set at a constant 30 FPS
-  videoComposition.frameDuration = CMTimeMake(1, 30);
+  videoComposition.sourceTrackIDForFrameTiming = videoTrack.trackID;
+  if (CMTIME_IS_VALID(videoTrack.minFrameDuration)) {
+    videoComposition.frameDuration = videoTrack.minFrameDuration;
+  } else {
+    NSLog(@"Warning: videoTrack.minFrameDuration for input video is invalid, please report this to "
+          @"https://github.com/flutter/flutter/issues with input video attached.");
+    videoComposition.frameDuration = CMTimeMake(1, 120);
```

Comment:
And why not 60? There is no good universal value. I think some displays have 120 Hz and this should not be ever used anyway.

Reply:
> And why not 60?

Because I would like the fallback to be conservative, and I think 30 is a reasonable baseline value for video. Which is why I proposed it at the beginning of the review.

> this should not be ever used anyway

If you believe this code path will never be used, I'm not sure why you have pushed back repeatedly against my suggestions for how we handle it.

Reply:
Hardcoded constant of 30 fps was what led to this PR in the first place. With 30 the worst case would be worse playback for videos with more fps, while something higher is more inclusive and worst case could be maybe performance related but I did not observe that even when I set 1000 or 1000000 here and it never produces more frames than are in source video. Actually my constant is conservative, it does not account for slow motion videos (in case it has actual frame rate for example 240 fps and not just 30 fps but it is slower, it can have 240 fps with instruction to play at 0.125 rate, in such case `frameDuration` need to be `CMTimeMake(1, 240)` and with 120 it would play at 15 fps). Also I think more "standard" is actually 24 fps (devised some 100 years ago), not 30, and 120 is divisible by 24.

Reply:
> Hardcoded constant of 30 fps was what led to this PR in the first place.

And this PR changes the use of the hard-coded constant from "always" to "cases where the frame duration can't be computed, of which we currently have no known examples", which constitutes a substantial improvement.

> With 30 the worst case would be worse playback for videos with more fps, while something higher is more inclusive and worst case could be maybe performance related

Yes, I am fully aware of that. I would prefer that cases we don't have a good understanding of and are actively reporting via logging as not handled correctly err on the side of worse playback rather than app performance problems.

For instance, we know from issue reports that people use this package to play audio files even though we do not recommend that, and I would rather that have worse (imaginary) video playback than performance problems if that's a case that can trigger this fallback.

> Also I think more "standard" is actually 24 fps (devised some 100 years ago), not 30, and 120 is divisible by 24.

For a fallback where we have no information about the video's framerate, I am more concerned with screen rate than film standards. Many common phone screen refresh rates are not divisible by 24.

Reply:
> Yes, I am fully aware of that. I would prefer that cases we don't have a good understanding of and are actively reporting via logging as not handled correctly err on the side of worse playback rather than app performance problems.

I do not think it would have worse performance with 120 vs 30 (I can think of a passive implementation where it does not even require some sort of timer which calls some function at that frequency). But even if it had a little impact there is still this `sourceTrackIDForFrameTiming` so it should mostly use track for timing anyway and just in those rarer occasions fallback to `frameDuration`.

> For instance, we know from issue reports that people use this package to play audio files even though we do not recommend that, and I would rather that have worse (imaginary) video playback than performance problems if that's a case that can trigger this fallback.

Yes, as I wrote audio tracks are the only ones for which I noticed that their `minFrameDuration` is invalid but video composition is used only when there is at least one video track (and specifically for the first video track). It would crash long before it could get there and use that 30 or 120 fps.

> For a fallback where we have no information about the video's framerate, I am more concerned with screen rate than film standards. Many common phone screen refresh rates are not divisible by 24.

Btw on macos the display refresh rate can be set to 24 and combination 120 in `frameDuration` with 24 fps video makes for smooth playback.

So why to not use 24 instead of 30? Or maybe just put here `CMTimeMake(1, 1)` for these imaginary videos?

Reply:
> for these imaginary videos?

Again: If you believe this code path will never be used, I'm not sure why you have pushed back repeatedly against my suggestions for how we handle it.

Reply:
I do not know whether it will be ever used or not. As I wrote, 120 seems to me a better choice than 30. Actually from some points of view 30 may be worse than even some ridiculous values like 10 where it will be at least very well noticeable than just barely noticeable but not quite right.

Reply:
> I do not know whether it will be ever used or not.

It is unconstructive to repeatedly assert that this case will "not be ever used" and is "imaginary" as a way to dismiss my position, and then argue the reverse when it comes to your position.

If you believe this cannot happen, then there is literally no downside to just making the change that I am asking for; if you do believe could potentially happen, then please stop asserting that it can't.

> As I wrote, 120 seems to me a better choice than 30.

Yes, I understand that, I just don't agree. I do not find the arguments you've presented compelling enough to warrant approving a behavioral change (relative to the version that is already in production) for a case we have no actual examples of, and thus no ability to evaluate concretely.

Reply:
When I mentioned "imaginary" I was referring to your "...have worse (**imaginary**) video..." before and it was just a reaction to what you wrote. I wrote "this should not be ever used" not "(will) not be ever used" and this does not mean that I "believe" it will certainly never happen (in such case I would write "this will not be ever used" as if it was literally dead code). I can think that chance is small but this is really unknown until I can test it on every existing video in the world.

So if I understand correctly your arguments are:

1. Better to use something that was here before as fallback due to behavioral change.
2. It should be "harmonic" with refresh rate, so nothing like 24 fps, to avoid resulting artifacts and worse playback.
3. It should be something high enough as to not cause worse playback than 30 fps would, so nothing less than 30 like 15 or 1.
4. It should not be too high due to performance concerns, even 60 is too much.

I would at least reconsider to use 60 fps here as it covers both 30 and 60 fps videos and although 30 fps is probably still more used, both are very popular and 60 just covers both of them. And even if setting 60 here would mean that video composition would try to check 2x more often whether there is new frame, consider that even with 30 here player plugin itself already needs to check for new frame at display refresh rate which is at least 60 Hz (seems there is possibility for even 30 Hz and lower but it is opt-in in [CADisplayLink](https://developer.apple.com/documentation/quartzcore/cadisplaylink/3875343-preferredframeraterange?language=objc)).

Reply:
> When I mentioned "imaginary" I was referring to your "...have worse (**imaginary**) video..." before and it was just a reaction to what you wrote.

The text of mine you are quoting is from a paragraph that was explicitly about *audio files*, where there *is* no video. Thus, the video playback in question would be imaginary.

Your use of "imaginary" was in reply to a later paragraph where I was talking about "fallback where we have no information about the video's framerate"; i.e., every case that will hit this line of code.

Those aren't the same cases, and it's important to distinguish between the specific case of audio files, and the general case (which we don't have a clear enumeration of) where we get no framerate information.

> I would at least reconsider to use 60 fps here

That does not satisfy point 1. If you can provide a concrete example of a video file that falls into this case, I'm happy to consider changing the value away from what we previously had since we will have at least one data point for what the impact of that change will be. If not, I don't want to make speculative changes about cases we don't have any information about.

Reply:
> Right, sorry; I was misremembering the logic used to set `videoComposition.frameDuration` logic as feeding into the display link.

So maybe this is the reason for your performance concerns? Now as `frameDuration` is not used for display link frequency, does this change the whole situation? To recapitulate as I wrote earlier, video output never gives more pixel buffers per second than is fps of video even with values like 1000 or million and I did not notice anything about performance with such values. Also when `sourceTrackIDForFrameTiming` is used then timing is mainly driven by track which even further minimizes possible (if any) negative impact of `frameDuration`. So there are several indices that indicate that there should not be any performance concerns.

Reply:
> So maybe this is the reason for your performance concerns? Now as `frameDuration` is not used for display link frequency, does this change the whole situation?

That certainly reduces my concerns somewhat. However:

> video output never gives more pixel buffers per second than is fps of video

What exactly does that mean in the context of a video where `frameDuration` is not set?

> even with values like 1000 or million and I did not notice anything about performance with such values

Did you test with any videos where `frameDuration` is not set, which is the only case where this code would come into play?

I continue not to see a compelling argument in favor of changing the behavior of a case we do not understand, and have no examples of. By definition, any such change is purely speculative and cannot be validated.